tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -3031,8 +3025,7 @@ class HLSLExternalSource : public ExternalSemaSource {
         &templateParamNamedDeclsCount);
     // }
 
-    SmallVector<hlsl::ParameterModifier, g_MaxIntrinsicParamCount> paramMods;
-    InitParamMods(intrinsic, paramMods);
+    auto paramMods = createParamMods(intrinsic);
```

Comment:
Revert back to passing the SmallVector as an argument. SmallVectors are kinda cool, but can be _really_ big. In this case the SmallVector's inline storage is big enough to store `g_MaxIntrinsicParamCount` elements. That means the copy on return is large and not just a `std::move` like you might have with a `std::vector`.

Reply:
Reverted. By default I assume return value optimization is in place, which is why I made this change as a slight readability improvement. Otherwise, this line of discussion would require to never return anything larger than `uintpr_t`. Admittedly, I have not checked if the extra copy actually was elided.

Reply:
Good to know.
note: just looked at the generated assembly in release, and the whole smallvector and createParamMods implem get's inlined, and looking at the debug build, it seems that we do get the copy elision